python/lsst/ip/diffim/detectAndMeasure.py

```diff
@@ -949,6 +982,101 @@ def calculateMetrics(self, science, difference, diaSources, kernelSources):
+            json={"visit_id": visit_id, "detector_id": detector_id, "diasources": dia_sources_json})
+
+        if sattle_output.status_code != 200:
+            # Check if the cache is missing the visit and retry once
+            if sattle_output.status_code == 404:
+                try:
+                    self.log.info('Visit not found in sattle cache, re-sending')
+
+                    visit_id = science.getInfo().getVisitInfo().id
+
+                    visit_date = Time(
+                        science.getInfo().getVisitInfo().getDate().toPython()).jd
+
+                    exposure_time_days = science.getInfo().getVisitInfo().getExposureTime() / 86400.0
+                    exposure_end_jd = visit_date + exposure_time_days / 2.0
+                    exposure_start_jd = visit_date - exposure_time_days / 2.0
+
+                    boresight_ra = science.getInfo().getVisitInfo().boresightRaDec[0].asDegrees()
+                    boresight_dec = science.getInfo().getVisitInfo().boresightRaDec[1].asDegrees()
+
+                    r = requests.put(
+                        f'{self.config.sattle_host}:{self.config.sattle_port}/visit_cache',
+                        json={"visit_id": visit_id,
+                              "exposure_start_mjd": exposure_start_jd,
+                              "exposure_end_mjd": exposure_end_jd,
+                              "boresight_ra": boresight_ra,
+                              "boresight_dec": boresight_dec,
+                              "historical": self.config.sattle_historical})
+
+                    if r.status_code != 200:
+                        raise RuntimeError(r.text)
+
+                    sattle_output = requests.put(
+                        f'{self.config.sattle_host}:{self.config.sattle_port}/'
+                        'diasource_allow_list',
+                        json={"visit_id": visit_id, "detector_id": detector_id,
+                              "diasources": dia_sources_json})
+
+                    # fail on any non-success
+                    if sattle_output.status_code != 200:
+                        raise RuntimeError(sattle_output.text)
+
+                except (requests.RequestException, ConnectionError) as e:
+                    raise RuntimeError(sattle_output.text) from e
+
+            else:
+                raise RuntimeError(sattle_output.text)
+
+        sattle_output_array = json.loads(sattle_output.content)
+
+        if sattle_output_array['allow_list'] == []:
+            self.log.warning('Sattle output array is empty, all sources removed')
+            diaSources = diaSources[0:0].copy(deep=True)
+        else:
+
+            allowed_ids = []
+            for source in diaSources:
+                if source['id'] in sattle_output_array['allow_list']:
+                    allowed_ids.append(True)
+                else:
+                    allowed_ids.append(False)
```

Comment:
```suggestion
            allow_set = set(sattle_output_array["allow_list"])
            allowed_ids = [source["id"] in allow_set for source in diaSources]
```
If `diaSources` were a simple list,
```suggestion
            allow_set = set(sattle_output_array["allow_list"])
            diaSources = [source if source["id"] in allow_set for source in diaSources]
```
would simplify things even more, but I assume it's not.

Reply:
`diaSources` are an `lsst.afw.table.SourceCatalog`, by the way.